### ISSUE TITLE
WebSocket closes automatically after 500ms

### ISSUE THREAD

**Issue by haitlahcen** [CleverCloud/clever-tools#296]
As stated [here](https://github.com/CleverCloud/clever-tools/blob/master/src/models/ws-stream.js#L70), when a `WebSocket` stream is opened, it is infinite. 

This is very problematic in case of network issues (i.e. the CLI will never exit). 

Maybe we should consider having a limit for the number of retry with an incremental timeout for the re-connection.

**Comment by clementd-fretlink** [comment]
While investigating on #294, we've been able to get a bit more information (by adding some logs in `ws-stream.js`). It took some time because we had to deploy a custom version with extra logs, and our CI environment makes it a bit difficult to get logs, but we got them eventually.

So, the current state of affairs is:
- the websocket closes "normally" (we get a `Normal Closure` event in the `close` event handler)
- judging by the timestamp, this event is received every 2s seconds. Since the reconnection timeout is 1.5s, it seems to indicate that the connection closes after 500ms.

This would expain the "not stopping" behaviour, since the websocket is closed ~75% of the time, it misses events.

With that in mind, I have two questions:
- do you have more information in the server logs?
- would getting events through SSE instead of websockets be feasible?

**Comment by hsablonniere** [comment]
@clementd-fretlink @haitlahcen Thanks for the details, it helps a lot.

**Comment by clementd-fretlink** [comment]
(I can submit a PR with the more detailed logs if you want)

**Comment by hsablonniere** [comment]
(I'm currently adding logs locally to get a better understanding of what's going on but please do, it can only help me a bit more)

**Comment by divarvel** [comment]
We're running one last test, to make sure the WS is not closed by the cleanup function of the observable (it's the only place `ws.close` is called, so…)

**Comment by divarvel** [comment]
Ok, so we are now sure that it isn't the client that triggers the close.
@hsablonniere could you describe what happens when the auth fails? Is the server closing the connection directly? Does it send a message in the socket describing the error?

**Comment by hsablonniere** [comment]
I tried with a random string for the auth to force a bad auth. I got this:

```
ENTER openStream(/*...*/)
ENTER openWebSocket(/*...*/)
ENTER new WebSocket(/*...*/)
[DEBUG] Websocket opened successfully: wss://api.clever-cloud.com/v2/events/event-socket
[DEBUG] Websocket closed.
[WARN] Websocket closed, reconnecting
CLEANUP CLOSE
```

**Comment by hsablonniere** [comment]
If the host is unreachable, the command will stop directly with an error.

**Comment by hsablonniere** [comment]
I also tried to shut off my network connexion while I was receiving events.

* Seems like no errors are detected
* pings are sent every 40s but without errors (I'm not sure we're supposed to ping heartbeats from the client)

If I put the network connexion back on, after a few secs, it works again without doing anything.

**Comment by hsablonniere** [comment]
My conclusion:

* I agree that we should not try to recreate the WebSocket client forever and I DO want to put some kind of timeout strategy
* I'm still lost on the exact cause that you both have regarding this problem

I'm not sure I fully understand what's happening on your side.

**Comment by clementd-fretlink** [comment]
We've reproduced outside the CI, it seems the connection is closed right after we send the auth message, so it really looks like an auth issue. It's strange because the HTTP calls don't have issues (and we push to git over https). Maybe logging the auth errors in the server, or sending back an error message in the socket before closing it could clear things up?

**Comment by hsablonniere** [comment]
OK I see.

So you mean with a given pair of token/secret, API calls and git push over HTTPS works but the oauth round trip over WS to get events fails, is this right?

I would assume it always fails or is it random?

I'm going to investigate on how the server part works...

I agree on the fact that the server should send a message regarding the failed auth.

**Comment by clementd-fretlink** [comment]
The WS auth always fail during a deployment. I think the first step would be to send a message, or _at the very least_ log an error in the server logs.

**Comment by divarvel** [comment]
But to properly solve the issue, we'll need the server to be more informative (right now we can't really know why the socket is closed).

**Comment by paulrbr-fl** [comment]
@hsablonniere I have finally been able to reproduce the behavior locally (not in our CI). By simply using the clever token/secret we had in our CI config. This token is pretty old (I would say 1year+ but I don't have the exact date).

It seems the token works for any synchronous commands (`clever status` works perfectly fine, `clever activity` works fine) but for asynchronous commands we get a never ending command (with `clever activity -f` or `clever deploy`) with the error described in the issue.

Clearly @divarvel 's proposition would help (server message during bad auth during the websocket)authentication).

**Comment by hsablonniere** [comment]
@paulrbr-fl Thanks. I was expecting something like that, I'm now investigating the details of why an old token could provoke this.

**Comment by paulrbr-fl** [comment]
@hsablonniere I have written to the support right now. Let us know if you need anything more.

**Comment by hsablonniere** [comment]
OK, so it seems to be related to missing permissions/rights with your token.
Please generate a new one.

I created an issue to improve the auth error behaviour so the CLI can do a better retry.

I will introduce a max retries to stop the endless WS. I think 5 is OK.

**Comment by divarvel** [comment]
Now that the client sends pings regularly, the connection won't be closed if there is no activity (that was the reason for endless automatic reconnection). Knowing where the error comes from is the most important IMO. Is using SSE instead of websocket still considered? That would be a nice improvement for the CLI as well as for other clients.

Thanks for your help :-)

**Comment by hsablonniere** [comment]
Yep, SSE is still in roadmap, not sure about the priority.

**Comment by hsablonniere** [comment]
Fixed by #300
